File: src/server/geo.ts
/** Great-circle helpers, shared by the rail graph and the position engine. */

export const EARTH_KM = 6371;

export const rad = (d: number): number => (d * Math.PI) / 180;
export const deg = (r: number): number => (r * 180) / Math.PI;

export interface LatLon {
  lat: number;
  lon: number;
}

export function haversine(a: LatLon, b: LatLon): number {
  const dLat = rad(b.lat - a.lat);
  const dLon = rad(b.lon - a.lon);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(rad(a.lat)) * Math.cos(rad(b.lat)) * Math.sin(dLon / 2) ** 2;
  return 2 * EARTH_KM * Math.asin(Math.sqrt(h));
}

export function haversineRaw(aLat: number, aLon: number, bLat: number, bLon: number): number {
  return haversine({ lat: aLat, lon: aLon }, { lat: bLat, lon: bLon });
}

/** Initial bearing a -> b, degrees clockwise from north. */
export function bearing(a: LatLon, b: LatLon): number {
  const dLon = rad(b.lon - a.lon);
  const y = Math.sin(dLon) * Math.cos(rad(b.lat));
  const x =
    Math.cos(rad(a.lat)) * Math.sin(rad(b.lat)) -
    Math.sin(rad(a.lat)) * Math.cos(rad(b.lat)) * Math.cos(dLon);
  return (deg(Math.atan2(y, x)) + 360) % 360;
}

/** Point at fraction f (0..1) along the great circle a -> b. */
export function greatCircle(a: LatLon, b: LatLon, f: number): LatLon {
  const d = haversine(a, b) / EARTH_KM;
  if (d < 1e-9) return { lat: a.lat, lon: a.lon };
  const A = Math.sin((1 - f) * d) / Math.sin(d);
  const B = Math.sin(f * d) / Math.sin(d);
  const x =
    A * Math.cos(rad(a.lat)) * Math.cos(rad(a.lon)) +
    B * Math.cos(rad(b.lat)) * Math.cos(rad(b.lon));
  const y =
    A * Math.cos(rad(a.lat)) * Math.sin(rad(a.lon)) +
    B * Math.cos(rad(b.lat)) * Math.sin(rad(b.lon));
  const z = A * Math.sin(rad(a.lat)) + B * Math.sin(rad(b.lat));
  return { lat: deg(Math.atan2(z, Math.hypot(x, y))), lon: deg(Math.atan2(y, x)) };
}

File: server/internal/geo/geo.go
// Package geo holds the great-circle helpers shared by the rail graph and the
// position engine.
//
// Distances are kilometres and bearings are degrees clockwise from north,
// matching the units the rest of the server speaks. Nothing here allocates:
// these run inside the routing inner loop, over a graph of 181 290 nodes.
package geo

import "math"

// EarthKm is the mean Earth radius used for every distance in the project.
// One value, so a path's length and a train's progress along it can never
// disagree by a rounding of the radius.
const EarthKm = 6371

// Point is a WGS84 coordinate.
type Point struct {
	Lat float64
	Lon float64
}

func rad(d float64) float64 { return d * math.Pi / 180 }
func deg(r float64) float64 { return r * 180 / math.Pi }

// Haversine returns the great-circle distance between two points, in km.
func Haversine(a, b Point) float64 {
	return HaversineAt(a.Lat, a.Lon, b.Lat, b.Lon)
}

// HaversineAt is Haversine on bare coordinates, for callers holding parallel
// arrays rather than structs — which, in the graph, is all of them.
func HaversineAt(aLat, aLon, bLat, bLon float64) float64 {
	dLat := rad(bLat - aLat)
	dLon := rad(bLon - aLon)
	sinLat := math.Sin(dLat / 2)
	sinLon := math.Sin(dLon / 2)
	h := sinLat*sinLat + math.Cos(rad(aLat))*math.Cos(rad(bLat))*sinLon*sinLon
	return 2 * EarthKm * math.Asin(math.Sqrt(h))
}

// Bearing returns the initial bearing from a to b, in degrees clockwise from
// north.
func Bearing(a, b Point) float64 {
	dLon := rad(b.Lon - a.Lon)
	y := math.Sin(dLon) * math.Cos(rad(b.Lat))
	x := math.Cos(rad(a.Lat))*math.Sin(rad(b.Lat)) -
		math.Sin(rad(a.Lat))*math.Cos(rad(b.Lat))*math.Cos(dLon)
	return math.Mod(deg(math.Atan2(y, x))+360, 360)
}

// GreatCircle returns the point at fraction f along the great circle from a
// to b. f is not clamped: the caller decides whether overshooting is meaningful.
func GreatCircle(a, b Point, f float64) Point {
	d := Haversine(a, b) / EarthKm
	if d < 1e-9 {
		return a
	}
	sinD := math.Sin(d)
	ca := math.Sin((1-f)*d) / sinD
	cb := math.Sin(f*d) / sinD
	x := ca*math.Cos(rad(a.Lat))*math.Cos(rad(a.Lon)) + cb*math.Cos(rad(b.Lat))*math.Cos(rad(b.Lon))
	y := ca*math.Cos(rad(a.Lat))*math.Sin(rad(a.Lon)) + cb*math.Cos(rad(b.Lat))*math.Sin(rad(b.Lon))
	z := ca*math.Sin(rad(a.Lat)) + cb*math.Sin(rad(b.Lat))
	return Point{Lat: deg(math.Atan2(z, math.Hypot(x, y))), Lon: deg(math.Atan2(y, x))}
}
